---
layout: page
title: "Computational analysis of sequencing based diagnostic testing data"
---

## Overview

Several groups have been developing large scale testing for SARS-CoV-2 based on high-throughput sequencing:

- [Jones et al., 2020](https://www.notion.so/Octant-SwabSeq-Testing-9eb80e793d7e46348038aa80a5a901fd) (SwabSeq)
- [Schmid-Burgk et al., 2020](https://www.biorxiv.org/content/10.1101/2020.04.06.025635v1.abstract) (LAMP-Seq)
- [Hossain et al., 2020](https://docs.google.com/document/d/1kP2w_uTMSep2UxTCOnUhh1TMCjWvHEY0sUUpkJHPYV4/edit)

These methods have the potential to massively scale the extent of SARS-CoV-2 testing at low-cost and with assays that are highly sensitive. They can also be easily modified to perform diagnostic testing for other viruses, for bacteria, and for genotyping.

The sequence data that will be generated by the proposed methods consists of two data types:

- (synthetic) sequence barcodes, each of which is associated to a sample.
- (biological) sequences associated with genes. Genes may include viral genes, control genes, or spike-ins.

Pre-processing of the data involves several steps:

- Error correction of the barcode sequences. This is necessary in order to account for sequence errors that may have been introduced during sequencing.
- Association of the biological sequences to their genes of origin.
- Collation of sequences associated to a single sample in order to count the number of times different genes have been observed.

Following pre-processing of the data, analysis must be performed to determine the identity of samples containing SARS-CoV-2 (positives), and those absent the virus (negatives). This involves setting thresholds based on the observed viral counts in the context of spike-in and control counts.

## kallisto bustools workflow

To analyze diagnostic testing sequence data we have adapted kallisto and bustools to perform all the necessary processing steps. The workflow can be run with just one command using `kb`, but "under the hood" it is organized as follows:

- An "index" (compressed data structure) is constructed from the gene sequences that will be counted in the diagnostic test. These gene sequences may consist of human sequence, viral sequence, or synthetic sequence. The index is constructed with kallisto using the `kallisto index` option. The input consists of the gene sequences in FASTQ format.
- Raw sequence data in FASTQ format is first converted into BUS file format using kallisto. 
  - kallisto is called with an option specific to the diagnostic technology. For example, for SwabSeq, kallisto is called with `kallisto -x SwabSeq`. This specifies the location of the sample barcodes in the read files, along with the location of the biological sequence. It is straightforward to add technology (-x) options to kallisto. 
  - kallisto process the FASTQ records one at a time, and for each records the sample barcode in the first column of the resultant BUS file. In the case of SwabSeq, the barcode sequence is split between reads, and kallisto automatically handles this (thanks to the specification with the -x SwabSeq option in kallisto). When processing a record, kallisto also aligns the biological sequence to the genes using the index, and results of the alignment are stored in the third column of the BUS file. The second column of the BUS file is unused for this application and is filled with a dummy "A". More detail on the alignment procedure is provided below.
  - Once kalliso has finished processing the reads, the resultant BUS file contains a table where each record corresponds to a sequence obtained from a sample, and where the record contains information about its sample barcode and a gene (which the biological sequence corresponding to that record aligned to). This BUS file is ready to be processed with various bustools programs.
- `bustools sort` is run to sort the BUS file by barcode sequence. This step is key to ensuring that all subsequent bustools commands run in constant memory, and time that is linear in the number of BUS records.
- `bustools correct` is now used to correct errors in the barcodes in the BUS file. This is done using a "whitelist" of barcodes from a sample sheet describing the diagnostic experiment. The `bustools correct` command will correct all barcodes within Hamming distance 1 of a whitelist barcode.
  - Alternatively, the `bustools whitelist` command can be run before `bustools correct` to automatically identify valid barcodes (using the total counts associated with each barcode). The reusltant whitelist can then be used with the `bustools correct` command to error correct the remaining barcodes.
- `bustools sort` is run again. The purpose of this sort command is to collate the counts of genes associated to each barcode sequence. These counts are recorded in a fourth column of the BUS file, indicating multiplcity of the record.
- `bustools count` is used to produce a human readable table of sample barcode sequences along with the counts of genes detected for each sample.

An example Google Colaboratory demonstrating the workflow for SwabSeq is available [here](https://github.com/pachterlab/BLGKP_2020/blob/master/notebooks/diagnostic.ipynb).



{% include JB/setup %}


